File: backend/app/core/sentry.py
"""Optional backend Sentry initialization and event filtering."""

from __future__ import annotations

from collections.abc import Mapping, MutableMapping
import logging
import re
from typing import Any

from .config import Settings

LOGGER = logging.getLogger(__name__)

_FILTERED_VALUE = "[Filtered]"
_NOISE_PATHS = {"/health"}
_SENTRY_INITIALIZED = False
_COMMON_EVENT_TAGS: dict[str, str] = {}
_SERVICE_NAME = "backend-api"
_SERVICE_SURFACE = "fastapi-api"
_SENSITIVE_KEY_PARTS = (
    "authorization",
    "cookie",
    "token",
    "secret",
    "password",
    "apikey",
    "jwt",
)
_POLICY_TEXT_KEYS = {
    "body",
    "requestbody",
    "termstext",
    "submittedtext",
    "normalizedtext",
    "normalizedtextbody",
    "rawtextbody",
    "capturedtext",
    "policytext",
    "rawpolicytext",
    "fullpolicytext",
    "rawinputexcerpt",
}


def init_sentry(*, settings: Settings) -> bool:
    """Initialize Sentry when a DSN is configured.

    The backend must remain safe in local and unset-DSN environments, so this
    function returns without side effects when Sentry is not configured.
    """

    global _SENTRY_INITIALIZED

    if _SENTRY_INITIALIZED:
        return True

    dsn = settings.sentry_dsn.strip()
    if not dsn:
        return False

    try:
        import sentry_sdk
        from sentry_sdk.integrations.fastapi import FastApiIntegration
    except ImportError:
        LOGGER.warning(
            "Sentry DSN is configured but sentry-sdk is not installed; skipping backend Sentry initialization."
        )
        return False

    release = settings.sentry_release.strip()
    environment = settings.effective_sentry_environment
    _COMMON_EVENT_TAGS.clear()
    _COMMON_EVENT_TAGS.update(
        {
            "service": _SERVICE_NAME,
            "service_surface": _SERVICE_SURFACE,
            "deployment_environment": environment,
        }
    )
    if release:
        _COMMON_EVENT_TAGS["release"] = release

    sentry_sdk.init(
        dsn=dsn,
        environment=environment,
        release=release or None,
        traces_sample_rate=_normalize_sample_rate(settings.sentry_traces_sample_rate),
        send_default_pii=False,
        max_request_body_size="never",
        before_send=_before_send,
        before_send_transaction=_before_send_transaction,
        integrations=[FastApiIntegration(transaction_style="url")],
    )
    for tag_name, tag_value in _COMMON_EVENT_TAGS.items():
        sentry_sdk.set_tag(tag_name, tag_value)
    _SENTRY_INITIALIZED = True
    return True


def _before_send(event: Any, hint: dict[str, Any]) -> dict[str, Any] | None:
    """Filter low-value events and scrub request details before sending."""

    _ = hint
    if not isinstance(event, MutableMapping):
        return None

    if _is_noise_event(event):
        return None

    _apply_common_tags(event)
    _scrub_request_data(event)
    _scrub_event_payload(event)
    return event


def _before_send_transaction(event: Any, hint: dict[str, Any]) -> dict[str, Any] | None:
    """Drop noisy transactions such as health checks."""

    _ = hint
    if not isinstance(event, MutableMapping):
        return None

    if _is_noise_event(event):
        return None
    _apply_common_tags(event)
    _scrub_event_payload(event)
    return event


def _is_noise_event(event: object) -> bool:
    if not isinstance(event, Mapping):
        return False

    transaction_name = str(event.get("transaction") or "").strip()
    if transaction_name in _NOISE_PATHS or transaction_name.endswith("/health"):
        return True

    request_data = event.get("request")
    if not isinstance(request_data, Mapping):
        return False

    request_url = str(request_data.get("url") or "").strip()
    if not request_url:
        return False

    return any(request_url.endswith(path) or f"{path}?" in request_url for path in _NOISE_PATHS)


def _scrub_request_data(event: MutableMapping[str, Any]) -> None:
    request_data = event.get("request")
    if not isinstance(request_data, MutableMapping):
        return

    headers = request_data.get("headers")
    if isinstance(headers, MutableMapping):
        for header_name in tuple(headers.keys()):
            if str(header_name).strip().lower() in {
                "authorization",
                "cookie",
                "set-cookie",
                "x-api-key",
            }:
                headers[header_name] = _FILTERED_VALUE

    for field_name in ("data", "cookies"):
        if field_name in request_data:
            request_data[field_name] = _FILTERED_VALUE


def _apply_common_tags(event: MutableMapping[str, Any]) -> None:
    tags = event.get("tags")
    if not isinstance(tags, MutableMapping):
        tags = {}
        event["tags"] = tags

    for tag_name, tag_value in _COMMON_EVENT_TAGS.items():
        tags.setdefault(tag_name, tag_value)


def _scrub_event_payload(event: MutableMapping[str, Any]) -> None:
    _scrub_mapping(event)


def _scrub_mapping(payload: MutableMapping[str, Any]) -> None:
    for key, value in tuple(payload.items()):
        if _should_filter_key(key):
            payload[key] = _FILTERED_VALUE
            continue

        if isinstance(value, MutableMapping):
            _scrub_mapping(value)
            continue

        if isinstance(value, list):
            _scrub_sequence(value)


def _scrub_sequence(values: list[Any]) -> None:
    for index, value in enumerate(values):
        if isinstance(value, MutableMapping):
            _scrub_mapping(value)
            continue

        if isinstance(value, list):
            _scrub_sequence(value)
            continue

        if isinstance(value, str) and _looks_like_bearer_token(value):
            values[index] = _FILTERED_VALUE


def _should_filter_key(key: object) -> bool:
    normalized_key = _normalize_key(str(key))
    if normalized_key in _POLICY_TEXT_KEYS:
        return True
    return any(part in normalized_key for part in _SENSITIVE_KEY_PARTS)


def _normalize_key(key: str) -> str:
    return re.sub(r"[^a-z0-9]+", "", key.lower())


def _looks_like_bearer_token(value: str) -> bool:
    return value.strip().lower().startswith("bearer ")


def _normalize_sample_rate(sample_rate: float) -> float:
    if sample_rate < 0:
        return 0.0
    if sample_rate > 1:
        return 1.0
    return sample_rate

File: frontend/src/observability/sentry.ts
import * as Sentry from '@sentry/react';

const LOCAL_API_TRACE_TARGET = /^https?:\/\/(localhost|127\.0\.0\.1)(:\d+)?\/api\//;
const EDGE_API_TRACE_TARGET = /^\/api\//;
const FILTERED_VALUE = '[Filtered]';
const SERVICE_NAME = 'frontend-dashboard';
const SERVICE_SURFACE = 'react-dashboard';
const SENSITIVE_KEY_PARTS = [
  'authorization',
  'cookie',
  'token',
  'secret',
  'password',
  'apikey',
  'jwt',
];
const POLICY_TEXT_KEYS = new Set([
  'body',
  'requestbody',
  'termstext',
  'submittedtext',
  'normalizedtext',
  'normalizedtextbody',
  'rawtextbody',
  'capturedtext',
  'policytext',
  'rawpolicytext',
  'fullpolicytext',
  'rawinputexcerpt',
]);

function readEnvValue(value: string | undefined): string {
  return value?.trim() ?? '';
}

function resolveSentryEnvironment(): string {
  const configured = readEnvValue(import.meta.env.VITE_SENTRY_ENVIRONMENT);
  if (configured) {
    return configured;
  }

  return import.meta.env.MODE;
}

function resolveSentryTracesSampleRate(): number {
  const configured = readEnvValue(import.meta.env.VITE_SENTRY_TRACES_SAMPLE_RATE);
  if (!configured) {
    return 0;
  }

  const parsed = Number(configured);
  if (!Number.isFinite(parsed)) {
    return 0;
  }

  return Math.min(1, Math.max(0, parsed));
}

function buildCommonTags(
  environment: string,
  release: string,
): Record<string, string> {
  const tags: Record<string, string> = {
    service: SERVICE_NAME,
    service_surface: SERVICE_SURFACE,
    deployment_environment: environment,
  };

  if (release) {
    tags.release = release;
  }

  return tags;
}

function applyCommonTags(
  event: Sentry.Event,
  tags: Record<string, string>,
): Sentry.Event {
  event.tags = {
    ...tags,
    ...(event.tags ?? {}),
  };
  return event;
}

function scrubEvent(event: Sentry.Event): Sentry.Event {
  return scrubUnknown(event) as Sentry.Event;
}

function scrubBreadcrumb(breadcrumb: Sentry.Breadcrumb): Sentry.Breadcrumb {
  return scrubUnknown(breadcrumb) as Sentry.Breadcrumb;
}

function scrubUnknown(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map((item) => scrubUnknown(item));
  }

  if (!value || typeof value !== 'object') {
    if (typeof value === 'string' && value.trim().toLowerCase().startsWith('bearer ')) {
      return FILTERED_VALUE;
    }
    return value;
  }

  const scrubbedEntries = Object.entries(value).map(([key, entryValue]) => {
    if (shouldFilterKey(key)) {
      return [key, FILTERED_VALUE];
    }
    return [key, scrubUnknown(entryValue)];
  });

  return Object.fromEntries(scrubbedEntries);
}

function shouldFilterKey(key: string): boolean {
  const normalizedKey = key.toLowerCase().replace(/[^a-z0-9]+/g, '');
  if (POLICY_TEXT_KEYS.has(normalizedKey)) {
    return true;
  }
  return SENSITIVE_KEY_PARTS.some((part) => normalizedKey.includes(part));
}

export function initDashboardSentry(): boolean {
  const dsn = readEnvValue(import.meta.env.VITE_SENTRY_DSN);
  if (!dsn || typeof window === 'undefined') {
    return false;
  }
  const environment = resolveSentryEnvironment();
  const release = readEnvValue(import.meta.env.VITE_SENTRY_RELEASE);
  const commonTags = buildCommonTags(environment, release);

  Sentry.init({
    dsn,
    enabled: true,
    environment,
    release: release || undefined,
    tracesSampleRate: resolveSentryTracesSampleRate(),
    integrations: [Sentry.browserTracingIntegration()],
    tracePropagationTargets: [EDGE_API_TRACE_TARGET, LOCAL_API_TRACE_TARGET],
    sendDefaultPii: false,
    beforeSend(event) {
      return applyCommonTags(scrubEvent(event), commonTags);
    },
    beforeSendTransaction(event) {
      return applyCommonTags(scrubEvent(event), commonTags);
    },
    beforeBreadcrumb(breadcrumb) {
      return scrubBreadcrumb(breadcrumb);
    },
  });
  for (const [tagName, tagValue] of Object.entries(commonTags)) {
    Sentry.setTag(tagName, tagValue);
  }
  return true;
}
